transformation_test/transformation_test.go

```diff
@@ -0,0 +1,217 @@
+package transformation_test
+
+import (
+	"bytes"
+	"context"
+	"embed"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+	"github.com/goccy/go-yaml"
+	"github.com/google/go-cmp/cmp"
+	"gotest.tools/v3/golden"
+)
+
+const (
+	flbMainConf          = "fluent_bit_main.conf"
+	flbParserConf        = "fluent_bit_parser.conf"
+	transformationInput  = "input.log"
+	transformationOutput = "output_fluentbit.yaml"
+	flbTag               = "transformation_test"
+)
+
+var (
+	flbPath = flag.String("flb", os.Getenv("FLB"), "Fluent-bit path")
+)
+
+//go:embed testdata
+var testdataDir embed.FS
+
+type transformationTest []loggingProcessor
+type loggingProcessor struct {
+	confgenerator.LoggingProcessor
+}
+
+func (l *loggingProcessor) UnmarshalYAML(ctx context.Context, unmarshal func(interface{}) error) error {
+	return confgenerator.LoggingProcessorTypes.UnmarshalComponentYaml(ctx, &l.LoggingProcessor, unmarshal)
+}
+
+func TestTransformationTests(t *testing.T) {
+	ctx := context.Background()
+	if len(*flbPath) == 0 {
+		t.Skip("--flb not supplied")
+	}
+
+	allTests, err := testdataDir.ReadDir("testdata")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, dir := range allTests {
+		dir := dir
+		t.Run(dir.Name(), func(t *testing.T) {
+			t.Parallel()
+			if !dir.IsDir() {
+				t.Fatal("testdata folder must only contain folders")
+			}
+
+			testStartTime := time.Now()
+			// Unmarshal transformation_config.yaml
+			var transformationConfig transformationTest
+			transformationConfig, err = readTransformationConfig(dir.Name())
+			if err != nil {
+				t.Fatal("failed to unmarshal config:", err)
+			}
+
+			// Generate config files
+			var genFiles map[string]string
+			genFiles, err = generateFluentBitConfigs(ctx, transformationConfig, dir.Name())
+
+			// Write config files in temp directory
+			tempPath := t.TempDir()
+			for k, v := range genFiles {
+				err := confgenerator.WriteConfigFile([]byte(v), filepath.Join(tempPath, k))
+
+				if err != nil {
+					t.Fatal(err)
+				}
+				t.Logf("generated file %q\n%s", k, v)
+			}
+
+			// Start Fluent-bit
+			arg := fmt.Sprintf("--config=%s --parser=%s -v", filepath.Join(tempPath, flbMainConf), filepath.Join(filepath.Join(tempPath, flbParserConf)))
+			cmd := exec.Command(fmt.Sprintf("%s/fluent-bit", *flbPath), strings.Split(arg, " ")...)
+
+			var stdout, stderr io.ReadCloser
+			stdout, err = cmd.StdoutPipe()
+			if err != nil {
+				t.Fatal("stdout pipe failure", err)
+			}
+			stderr, err = cmd.StderrPipe()
+			if err != nil {
+				t.Fatal("stderr pipe failure", err)
+			}
+
+			if err := cmd.Start(); err != nil {
+				t.Fatal("Failed to start command:", err)
+			}
+
+			// read stderr
+			slurp, _ := io.ReadAll(stderr)
```

Comment:
I think you need to actually read stderr and stdout in parallel; pipes only have a limited amount of buffer space, and if fluent-bit tries to write more output than will fit in the buffer it will block until the pipe is read, which will cause a deadlock. You can solve this with goroutines!